arbnode/espresso_caff_node.go

```diff
@@ -214,21 +284,95 @@ func NewEspressoCaffNode(
 
 	stateChecker := NewStateChecker(
 		configFetcher().StateChecker,
-		httpPort,
+		stack.Config().HTTPPort,
 		fatalErrChan,
 	)
 
+	// Create a new EspressoKeyManager
+	// Get the EspressoTEEVerifier address from SequencerInbox contract
+	configAddress := configFetcher().EspressoTEEVerifierAddr
+	var espressoTEEVerifierAddress common.Address
+	// parse the espresso tee verifier address from config if it exists, otherwise read from the sequencerInbox
+	// Eventually we should only read from the SequencerInbox
+	if configAddress != "" && common.IsHexAddress(configAddress) {
+		espressoTEEVerifierAddress = common.HexToAddress(configAddress)
+	} else {
+		espressoTEEVerifierAddress, err = sequencerInbox.con.EspressoTEEVerifier(&bind.CallOpts{})
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to get EspressoTEEVerifier address: %w", err)
+	}
+	espressoTEEVerifier, err := espressogen.NewIEspressoTEEVerifier(espressoTEEVerifierAddress, l1Reader.Client())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get nitro verifier address: %w", err)
+	}
+	verifier := espressotee.NewEspressoTEEVerifier(espressoTEEVerifierAddress.Hex(), l1Reader.Client(), espressoTEEVerifierAddress)
+
+	var nitroVerifier espressotee.EspressoNitroTEEVerifierInterface
+	if teeType == espresso_key_manager.NITRO {
+		log.Info("setting up nitro verifier", "tee type", teeType)
+		nitroVerifier, err = espresso_key_manager.SetupNitroVerifier(espressoTEEVerifier, l1Reader.Client(), espressotee.CaffNode)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var dataPoster *dataposter.DataPoster
+	var keyManager *espresso_key_manager.EspressoKeyManager
+	var snapshotHandler *EspressoSnapshotHandler
+	if teeType != espressotee.EMPTY {
+		if txOptsCaffNode == nil {
+			return nil, fmt.Errorf("non nil txOpts are required to run the Caff Node in a TEE")
+		}
+
+		dataPosterConfigFetcher := func() *dataposter.DataPosterConfig {
+			dpCfg := configFetcher().DataPoster
+			return &dpCfg
+		}
+
+		chainId, err := l1Reader.Client().ChainID(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get chain id: %w", err)
+		}
+
+		dataPoster, err := dataposter.NewDataPoster(ctx,
```

Comment:
Can you explain a bit why caff node has to post things to the parent chain?

Reply:
because it is also running in a TEE so its key needs to be registered in the contract

Reply:
Why its key needs to be registered? My understanding is that it is used to verify something. Is the caff supposed to post anything?

Reply:
When someone will read some data from the caff node in the future, they will check that the data is signed using the address of the caff node that got registered in the contract. That part of the workflow is not implemented yet but we will work on it in the future!